elegant/worm_data.py

```diff
@@ -441,6 +441,78 @@ def interpolate_feature(self, feature, age, age_feature='age'):
+
+            Parameters
+                filter_type - str corresponding to the filter to be made ('gaussian', 'uniform')
+                **filter_params - parameters for the specified filter; these include:
+
+                    window_length - required parameter for either filter (used for the gaussian to truncate it)
+                    sigma - used only for gaussian filters
+
+            Returns
+                The generated filter will have call signature filter(time, data)
+                where: (1) time has been previously centered so that t=0 represents
+                the current timepoint and positive values of time correspond to past data,
+                and (2) data contains values to smooth over.
+        '''
+
+        def gaussian_filter(time, data, window_length, sigma):
+            smoothing_mask = (time>=0) & (time <= window_length)
+            t = time[smoothing_mask]
+            return numpy.dot(1/numpy.exp(-t**2/(2*sigma)).sum() * numpy.exp(-t**2/(2*sigma)), data[smoothing_mask])
+        def uniform_filter(time, data, window_length):
+            smoothing_mask = (time>=0) & (time <= window_length)
+            return numpy.dot(smoothing_mask/smoothing_mask.sum(),data)
+
+        filters = {'gaussian': gaussian_filter,
+            'uniform': uniform_filter}
+
+        def smoothing_filter(time, data):
+            selected_filter = filters[filter_type](**filter_params)
+            return selected_filter(time,data)
+        return smoothing_filter
+
+
+    def smooth_feature(self, feature, filter_type, age_feature='age', min_age=-numpy.inf, max_age=numpy.inf, **filter_params):
+        """Smooth a feature with a prescribed filter
+
+            Parameters
+                feature - feature to filter over
+                filter_type - str corresponding to the filter to be applied ('gaussian', 'uniform'),
+                    or a callable with the call signature smoothing_filter(time,data)
+                age_feature: the name of the feature to use to determine the age
+                    window. Defaults to 'age', but other monotonic features, such as
+                    a time-left-alive 'ghost_age' could also be used. If this is a
+                    function, it will be called as age_feature(worm) to generate
+                    the ages to examine (see examples).
+                min_age, max_age - Limits between which to perform filtering;
+                    outside this range, no filtering is performed.
+                **filter_params - Additional parameters passed to make_smoothing_filter
+                    to create filter.
+
+            The smoothed feature is the same size as the original feature and is
+                stored in the td attribute "smoothed_(feature)".
+        """
+
+        if callable(filter_type):
+            smoothing_filter = filter_type
+        else:
+            smoothing_filter = make_smoothing_filter(filter_type, window_length, **filter_params)
```

Comment:
`window_length` isn't defined. That should just be one of the filter params now, no?